### PR TITLE
testing, testing

### LINKED ISSUE #97: [FEATURE] Testing/Mocking system
**Is your feature request related to a problem? Please describe.**
No but I can't test my bot's code

**Describe the solution you'd like**
Provide a good way to test the bot code like [Laravel](https://laravel.com/docs/8.x/mocking) and/or [Botman](https://botman.io/2.0/testing)

**Describe alternatives you've considered**
None.

**Additional context**
Example/Idea:
context

### PR DESCRIPTION
**Pull request by SergiX44**
closes #97

### CODE REVIEW

**Comment by Lukasss93** [comment]
add examples please 😈

**Comment by Lukasss93** [comment]
The changes in the last commit were inspired by https://botman.io/2.0/testing

How to:
immagine

Warning: this is WIP.

File: src/Testing/FakeNutgram.php [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    {
+        $this->mockHandler = $mockHandler;
+        return $this;
+    }
+
+    /**
+     * @param  mixed  $update
+     * @return $this
+     */
+    public function hears(mixed $update): static
```
Comment (Lukasss93) [inline src/Testing/FakeNutgram.php:108]: add faker

File: src/Testing/FakeNutgram.php [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        return $this;
+    }
+
+    /**
+     * @param  string  $method
+     * @param  mixed  $data
+     * @param  int  $index
+     * @return FakeNutgram
+     */
+    public function assertApiRequestContains(string $method, mixed $data, int $index = 0): self
```
Comment (Lukasss93) [inline src/Testing/FakeNutgram.php:169]: remove `string $method,`
just use `assertApiMethodCalled`

File: src/Testing/FakeNutgram.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+        return $this;
+    }
+
+    /**
+     * @param  string  $method
+     * @param  int  $times
+     * @return FakeNutgram
+     */
+    public function assertApiMethodCalled(string $method, int $times): self
```
Comment (Lukasss93) [inline src/Testing/FakeNutgram.php:146]: `int $times = 1`

**Comment by Lukasss93** [comment]
assertReplay expected data contains objects instead arrays!

immagine

immagine

File: src/Testing/Asserts.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    /**
+     * @param  array  $expected
+     * @param  int  $index
+     * @param  string|null  $forceMethod
+     * @return $this
+     */
+    public function assertMessage(array $expected, int $index = 0, ?string $forceMethod = null): self
+    {
+        $allowed = $forceMethod ?? [
```
Comment (SergiX44) [inline src/Testing/Asserts.php:93]: This list must be mantained, consider retrieving the methods on instantiation via reflection instead

File: src/Testing/Asserts.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+     * @param  int|null  $messageId
+     * @param  int  $index
+     * @return $this
+     */
+    public function assertDeletedMessage(?int $chatId = null, ?int $messageId = null, int $index = 0): self
+    {
+        if ($chatId !== null && $messageId !== null) {
+            return $this->assertReply('deleteMessage', [
+                'chat_id' => $chatId,
+                'message_id' => $messageId,
```
Comment (SergiX44) [inline src/Testing/Asserts.php:143]: Consider using the prefill from the $this->storedChat and user

**Review by SergiX44 — COMMENTED** [review]
@Lukasss93

File: src/Testing/Asserts.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+        /** @var Request $request */
+        [$request,] = array_values($reqRes);
+
+        PHPUnit::assertContains($request->getUri()->getPath(), $method, 'Method name not found');
+
+        if ($expected !== null) {
+            try {
+                $expected = json_decode(json_encode($expected), true);
+                $actual = json_decode((string)$request->getBody(), true, flags: JSON_THROW_ON_ERROR);
```
Comment (Lukasss93) [inline src/Testing/Asserts.php:78]: TODO: handle multipart request

Comment (Lukasss93) [reply src/Testing/Asserts.php:78]: unfixable 😥

**Review by Lukasss93 — APPROVED** [review]

**Pull request merged**